src/node.cc

```diff
@@ -1278,70 +1283,61 @@ void AddPromiseHook(v8::Isolate* isolate, promise_hook_func fn, void* arg) {
 }
 
 
-Local<Value> MakeCallback(Environment* env,
-                          Local<Value> recv,
-                          const Local<Function> callback,
-                          int argc,
-                          Local<Value> argv[]) {
+MaybeLocal<Value> MakeCallback(Environment* env,
+                               Local<Value> recv,
+                               const Local<Function> callback,
+                               int argc,
+                               Local<Value> argv[],
+                               double async_id,
+                               double trigger_id) {
   // If you hit this assertion, you forgot to enter the v8::Context first.
   CHECK_EQ(env->context(), env->isolate()->GetCurrentContext());
 
-  Local<Object> object, domain;
-  bool has_domain = false;
+  Local<Object> object;
 
   Environment::AsyncCallbackScope callback_scope(env);
+  bool disposed_domain = false;
 
   if (recv->IsObject()) {
     object = recv.As<Object>();
   }
 
   if (env->using_domains()) {
     CHECK(recv->IsObject());
-    Local<Value> domain_v = object->Get(env->domain_string());
-    has_domain = domain_v->IsObject();
-    if (has_domain) {
-      domain = domain_v.As<Object>();
-      if (domain->Get(env->disposed_string())->IsTrue())
-        return Undefined(env->isolate());
-    }
+    disposed_domain = DomainEnter(env, object);
+    if (disposed_domain) return Undefined(env->isolate());
   }
 
-  if (has_domain) {
-    Local<Value> enter_v = domain->Get(env->enter_string());
-    if (enter_v->IsFunction()) {
-      if (enter_v.As<Function>()->Call(domain, 0, nullptr).IsEmpty()) {
-        FatalError("node::MakeCallback",
-                   "domain enter callback threw, please report this");
-      }
-    }
-  }
+  MaybeLocal<Value> ret;
 
-  // TODO(trevnorris): Correct this once node::MakeCallback() support id and
-  // triggerId. Consider completely removing it until then so the async id can
-  // propagate through to the fatalException after hook calls.
-  AsyncHooks::ExecScope exec_scope(env, 0, 0);
+  {
+    if (trigger_id == -1)
+      trigger_id = async_id;
```

Comment:
> Since this means that `trigger_id` shouldn't be optional, perhaps we should also make it non-optional in `EmitAsyncInit`.

Yeah, right. I changed it to be non-optional, could you take a look at 6881fd6dcc74b7916e64ab2a8ded363e83796940?

I guess the question in the back of my head is: Why are we tracking and using the `trigger_id` after the `init` callback at all, if we [deliberately](https://github.com/nodejs/node/pull/13068) do not expose it? (Also: Is there a good place to ask that question, if not here?)